In [1]:
import pandas as pd
import numpy as np
import validators


In [2]:
tabsv = "data/off_dataset.tsv"

df_off = pd.read_csv(tabsv, sep="\t", dtype="unicode", low_memory=False)


In [70]:
len(df_off.columns)

162

In [3]:
print(f"""There is {len(df_off)} products in the dataset and\
 only {len(df_off[df_off['carbon-footprint_100g'].notna()])} contain their carbon footprint
This is not enough to draw an average
 """)

There is 320772 products in the dataset and only 268 contain their carbon footprint
This is not enough to draw an average
 


In [4]:
len(df_off[df_off["stores"].notna()])


51722

Among the 320722 products, only 51722 have non NaN "store" field. Building an application using the stores data won't be efficient in itself, but if the user of the potential application is invited to fill the store's field of a product, this could lead to interesting functionnalities

In [71]:
print(len(df_off[df_off["categories"].isna()]))
print(len(df_off[df_off["main_category"].isna()]))


236362
236406


236300+ products have no categories
pretty useless if a comparison must be done

In [72]:
print(len(df_off[df_off["nutrition_grade_fr"].isna()]))


99562


99562 products have no nutriscore attached to it

In [67]:
df_off["last_modified_t"].max()

'Suisse,France'

In [69]:
df_off["last_modified_t"]
len(df_off[pd.to_numeric(df_off['last_modified_t'], errors='coerce').notnull()])


320749

In [5]:
# Filters

def get_col_by_ending(df: pd.DataFrame, ending: str) -> list:
    """takes an ending and a df as parameters, returns all cols ending with parameter"""
    
    col_list = []

    for col in df.columns:
        if col.endswith(ending):
            col_list.append(col)

    return col_list



def filter_series(series: pd.Series) -> bool:
    """
    set of filter : if not url -> False, if not proper URL
    -> False, no name -> False, no nutriscore -> False
    if no category : False
    """
    if pd.isna(series["url"]):
        return False
    if not validators.url(series["url"]):
        return False
    if pd.isna(series["product_name"]):
        return False
    if pd.isna(series["nutrition_grade_fr"]):
        return False
    if pd.isna(series["categories"]) and pd.isna(series["main_category"]):
        return False
    
    return True


def filter_creator(creator: str) -> bool:
    pass


def ecolo_rate(series: pd.Series) -> dict:
    """
    returns if a products contains palm oil and/or if its labelled as bio
    0 -> unknown , False -> No, True -> Yes
    """
    ecolo_dict = {
        "palm_oil": 0,
        "bio": 0,
    }
    pass


def aberation_cleaner(series: pd.Series, hundred_g_list: list):
    """if a non na col of a 100g is > 100 or < 0, returns false"""

    for col in hundred_g_list:
        if pd.notna(series[col]):
            if float(series[col]) > 100 or float(series[col]) < 0:
                return False


def copy_cleaner(df: pd.DataFrame, subset: str):
    pass


In [56]:
%%time

hundred_g_list = get_col_by_ending(df_off, "100g")


CPU times: user 93 µs, sys: 0 ns, total: 93 µs
Wall time: 96.1 µs


In [57]:
%%time

sec_hundred_g_list = [col for col in df_off.columns if col.endswith("100g")]


CPU times: user 87 µs, sys: 1e+03 ns, total: 88 µs
Wall time: 91.1 µs


In [53]:
hundred_g_list == sec_hundred_g_list


True

In [7]:
print("products", len(df_off))
print("unique products", len(df_off["code"].unique()))


products 320772
unique products 320750


In [8]:
cc1 = "Coca Cola 50Cl"
cc2 = "Coca Cola 500ml"
cc3 = "Coca-Cola 50 Cl"
cc4 = "CocaCola 50Cl"
cc5 = "Coca-Cola 500ml"
cc6 = "Coca Cola .50L"
cc7 = "Coca Cola0,5L"
cc8 = "Coca Cola.5L"
cc9 = " cocacola 50cl"
cc10 = "coke cola 50cl" #  This exemple will not work

ccl = [cc1, cc2, cc3, cc4, cc5, cc6, cc7, cc8, cc9, cc10]


In [9]:
def detect_name_copy(analyzed_name: str, original_name: str) -> bool:
    pass

def parse_data(product_name: str):
    pass

def ean_thirteen_filter(df: pd.DataFrame):

    df.drop_duplicates(subset="code", keep="first")
    
    

In [10]:
converting_table_1_litre = {
    "l": 1,
    "cl": 100,
    "dl": 10,
    "ml": 1000
}

converting_table_1_gram = {
    "mg": 1000,
    "kg": 0.1
}


In [43]:
def detect_unit(product_name: str):
    pass

def convert_unit(product_name: str, subunit: str, conversion_table: dict):
    pass


In [44]:
%%time

columns_list = [col for col in df_off.columns if col not in hundred_g_list]


CPU times: user 237 µs, sys: 1e+03 ns, total: 238 µs
Wall time: 241 µs


In [45]:
%%time

columns_set = list(set(df_off.columns) - set(hundred_g_list))


CPU times: user 70 µs, sys: 1 µs, total: 71 µs
Wall time: 73.9 µs


In [50]:

for col in columns_list:
    if col not in columns_set:
        print("False")

for col in columns_set:
    if col not in columns_list:
        print("False")


['first_packaging_code_geo',
 'ingredients_from_palm_oil_tags',
 'allergens',
 'states_fr',
 'manufacturing_places_tags',
 'nutrition_grade_uk',
 'traces',
 'labels_fr',
 'purchase_places',
 'countries',
 'categories_tags',
 'countries_tags',
 'no_nutriments',
 'ingredients_that_may_be_from_palm_oil_n',
 'labels_tags',
 'origins_tags',
 'additives_tags',
 'pnns_groups_1',
 'brands',
 'pnns_groups_2',
 'brands_tags',
 'states',
 'origins',
 'creator',
 'stores',
 'ingredients_that_may_be_from_palm_oil',
 'traces_tags',
 'emb_codes',
 'categories',
 'labels',
 'allergens_fr',
 'cities_tags',
 'nutrition_grade_fr',
 'traces_fr',
 'created_datetime',
 'categories_fr',
 'last_modified_datetime',
 'manufacturing_places',
 'serving_size',
 'additives_n',
 'main_category_fr',
 'cities',
 'created_t',
 'additives',
 'generic_name',
 'states_tags',
 'image_url',
 'countries_fr',
 'packaging',
 'main_category',
 'packaging_tags',
 'quantity',
 'ingredients_from_palm_oil_n',
 'ingredients_from_pal